### PR TITLE
feat: use cli ls extension [IDE-76]

### PR DESCRIPTION
**Pull request by ShawkyZ**
### Description
* Remove LS Path from configuration and settings.
* Download CLI instead of LS.
* Run LS from CLI LS extension.
* Added CLI Release Channels.
* Added CLI base download URL. 

### Checklist

- [x] Tests added and all succeed
- [x] Linted
- [x] CHANGELOG.md updated
- [ ] README.md updated, if user-facing

### Screenshots / GIFs

_Visuals that may help the reviewer. Please add screenshots for any UI change. GIFs are most welcome!_

### CODE REVIEW

File: src/snyk/common/services/downloadService.ts [hunk 1]
```diff
@@ -2,39 +2,32 @@ import { ReplaySubject } from 'rxjs';
...
+    private readonly lsApi: IStaticCliApi,
     readonly window: IVSCodeWindow,
     private readonly logger: ILog,
     downloader?: Downloader,
   ) {
-    this.downloader = downloader ?? new Downloader(configuration, lsApi, window, logger);
+    this.downloader = downloader ?? new Downloader(configuration, lsApi, window, logger, this.extensionContext);
   }
 
   async downloadOrUpdate(): Promise<boolean> {
```
Comment (ShawkyZ) [inline src/snyk/common/services/downloadService.ts:29]: TODO: call this when a new manual scan is triggered

File: src/snyk/common/watchers/configurationWatcher.ts [hunk 2]
```diff
@@ -51,9 +52,13 @@ class ConfigurationWatcher implements IWatcher {
       await extension.contextService.setContext(SNYK_CONTEXT.LOGGEDIN, false);
       await extension.contextService.setContext(SNYK_CONTEXT.AUTHENTICATION_METHOD_CHANGED, true);
       return extension.viewManagerService.refreshAllViews();
-    } else if (key === ADVANCED_CUSTOM_LS_PATH) {
+    } else if (key === ADVANCED_CLI_PATH) {
       // Language Server client must sync config changes before we can restart
       return _.debounce(() => extension.restartLanguageServer(), DEFAULT_LS_DEBOUNCE_INTERVAL)();
+    } else if(key === ADVANCED_CLI_RELEASE_CHANNEL) {
+      await extension.stopLanguageServer();
```
Comment (ShawkyZ) [inline src/snyk/common/watchers/configurationWatcher.ts:59]: remove stopLanguageserver from here after it's implemented in the cli downloader

File: src/snyk/cli/cliExecutable.ts [hunk 3]
```diff
@@ -1,37 +1,69 @@
...
+import { Platform } from '../common/platform';
 
-// TODO: This file is to be removed in VS Code + Language Server feature cleanup. We need to ensure all users have migrated to use CLI path that's set by the language server.
 export class CliExecutable {
-  // If values updated, `.vscodeignore` to be changed.
   public static filenameSuffixes: Record<CliSupportedPlatform, string> = {
     linux: 'snyk-linux',
-    win32: 'snyk-win.exe',
-    darwin: 'snyk-macos',
+    linux_alpine: 'snyk-alpine',
```
Comment (bastiandoetsch) [inline src/snyk/cli/cliExecutable.ts:10]: what about alpine-arm64?

File: src/snyk/cli/cliExecutable.ts [hunk 4]
```diff
@@ -1,37 +1,69 @@
...
+    macos: 'snyk-macos',
+    macos_arm64: 'snyk-macos-arm64',
+    windows: 'snyk-win.exe',
```
Comment (bastiandoetsch) [inline src/snyk/cli/cliExecutable.ts:13]: what about windows-arm64?

File: src/snyk/cli/cliExecutable.ts [hunk 5]
```diff
@@ -1,37 +1,69 @@
...
+    return this.filenameSuffixes[platform];
+  }
+
+  static async getCurrentWithArch(): Promise<CliSupportedPlatform> {
+    let platform = '';
+    const osName = Platform.getCurrent().toString().toLowerCase();
+    const archName = Platform.getArch().toLowerCase();
+    if (osName === 'linux') {
+      if (await this.isAlpine()) {
+        platform = 'linux_alpine';
```
Comment (bastiandoetsch) [inline src/snyk/cli/cliExecutable.ts:37]: are you sure it's an underscore here?

File: src/snyk/cli/cliExecutable.ts [hunk 6]
```diff
@@ -1,37 +1,69 @@
...
+      } else {
+        platform = 'linux';
+      }
+    } else if (osName === 'darwin') {
+      if (archName === 'arm64') {
+        platform = 'macos_arm64';
```
Comment (bastiandoetsch) [inline src/snyk/cli/cliExecutable.ts:43]: same here - would think it needs to be a dash. can we test that in integration tests on the different operating systems?

File: src/snyk/cli/cliExecutable.ts (hunk 5 above)
Comment (ShawkyZ) [reply src/snyk/cli/cliExecutable.ts:37]: yes it's the enum that we have defined in line 12

File: src/snyk/cli/cliExecutable.ts (hunk 3 above)
Comment (ShawkyZ) [reply src/snyk/cli/cliExecutable.ts:10]: agree.. we need to handle all these cases here.. 
Also port the changes to IntelliJ afterwards.

File: src/snyk/cli/cliExecutable.ts (hunk 4 above)
Comment (ShawkyZ) [reply src/snyk/cli/cliExecutable.ts:13]: agree.. we need to handle all these cases here..
Also port the changes to IntelliJ afterwards.

File: src/snyk/cli/staticCliApi.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  getFileName(platform: CliSupportedPlatform): string {
+    return CliExecutable.getFileName(platform);
+  }
+
+  async getLatestCliVersion(releaseChannel: string): Promise<string> {
+    let { data } = await axios.get<string>(
+      this.getLatestVersionDownloadUrl(releaseChannel),
+      await getAxiosConfig(this.workspace, this.configuration, this.logger),
```
Comment (bastiandoetsch) [inline src/snyk/cli/staticCliApi.ts:49]: Historically, we sometimes had problems with Axios & proxies. Is this initialized correctly (like before)?

File: src/snyk/cli/supportedPlatforms.ts [hunk 8]
```diff
@@ -1,2 +1,2 @@
-const SupportedCliPlatformsList = ['linux', 'win32', 'darwin'] as const;
+const SupportedCliPlatformsList = ['linux', 'linux_alpine', 'windows', 'macos', 'macos_arm64'] as const;
```
Comment (bastiandoetsch) [inline src/snyk/cli/supportedPlatforms.ts:1]: same here with the platforms

File: src/snyk/common/configuration/configuration.ts [hunk 9]
```diff
@@ -489,8 +525,16 @@ export class Configuration implements IConfiguration {
...
 
-  getCliPath(): string | undefined {
-    return this.workspace.getConfiguration<string>(CONFIGURATION_IDENTIFIER, this.getConfigName(ADVANCED_CLI_PATH));
+  async getCliPath(): Promise<string | undefined> {
+    let cliPath = this.workspace.getConfiguration<string>(
+      CONFIGURATION_IDENTIFIER,
+      this.getConfigName(ADVANCED_CLI_PATH),
+    );
+    if (!cliPath && extensionContext) {
+      cliPath = await CliExecutable.getPath(extensionContext.extensionPath);
```
Comment (bastiandoetsch) [inline src/snyk/common/configuration/configuration.ts:534]: this line should be superfluous - it's already above in setCliPath

File: src/snyk/common/configuration/configuration.ts [hunk 10]
```diff
@@ -287,7 +321,9 @@ export class Configuration implements IConfiguration {
   }
 
   async setCliPath(cliPath: string | undefined): Promise<void> {
-    if (!cliPath) return;
+    if (!cliPath && extensionContext) {
+      cliPath = await CliExecutable.getPath(extensionContext.extensionPath);
```
Comment (bastiandoetsch) [inline src/snyk/common/configuration/configuration.ts:325]: why do we set it to extensionPath? shouldn't it be a subfolder? Or include the filename?

File: src/snyk/common/download/downloader.ts [hunk 11]
```diff
@@ -1,53 +1,51 @@
...
-    const lsDir = path.dirname(lsPath);
-    mkdirSync(lsDir, { recursive: true });
-    if (await this.binaryExists(lsPath)) {
-      await this.deleteFileAtPath(lsPath);
+  private async getCliExecutable(platform: CliSupportedPlatform): Promise<CliExecutable | null> {
+    const cliPath = await CliExecutable.getPath(
+      this.extensionContext.extensionPath,
+      await this.configuration.getCliPath(),
+    );
+    if (await this.binaryExists(cliPath)) {
```
Comment (bastiandoetsch) [inline src/snyk/common/download/downloader.ts:43]: here, the cli path is used including the binary - above it is set to default to the extension path. This feels like a potential bug.

File: src/snyk/common/download/downloader.ts [hunk 12]
```diff
@@ -1,53 +1,51 @@
...
+      await this.deleteFileAtPath(cliPath);
```
Comment (bastiandoetsch) [inline src/snyk/common/download/downloader.ts:44]: We should probably only delete after having downloaded successfully, right?

File: src/snyk/common/download/downloader.ts [hunk 13]
```diff
@@ -1,53 +1,51 @@
...
     }
-
-    const lsVersion = (await this.lsApi.getMetadata()).version;
-    const sha256 = await this.lsApi.getSha256Checksum(lsPlatform);
-    const checksum = await this.downloadLs(lsPath, lsPlatform, sha256);
+    const lsVersion = await this.cliApi.getLatestCliVersion(this.configuration.getCliReleaseChannel());
```
Comment (bastiandoetsch) [inline src/snyk/common/download/downloader.ts:46]: nitpick: variable name

File: src/snyk/common/download/downloader.ts [hunk 14]
```diff
@@ -78,25 +76,25 @@ export class Downloader {
...
       token.onCancellationRequested(async () => {
         requestToken.cancel();
         this.logger.info(messages.downloadCanceled);
-        await this.deleteFileAtPath(lsPath);
+        await this.deleteFileAtPath(cliPath);
       });
 
       progress.report({ increment: 0 });
-      return await this.doDownload(requestToken, token, lsPath, request, hash, progress);
+      return await this.doDownload(requestToken, token, cliPath, request, hash, progress);
```
Comment (bastiandoetsch) [inline src/snyk/common/download/downloader.ts:97]: maybe use a temp file path here and only delete old file when everything is successful, sha256 is correct. Afterwards move the temp file to clipath?

Comment (bastiandoetsch) [reply src/snyk/common/download/downloader.ts:97]: e.g. in IntelliJ, we do it like this: https://github.com/snyk/snyk-intellij-plugin/blob/1adcc9ca1fe0f5ec56f8534a45d5c3e5d60ac267/src/main/kotlin/io/snyk/plugin/services/download/CliDownloader.kt#L41

File: src/snyk/common/languageServer/languageServer.ts
Comment (bastiandoetsch) [inline src/snyk/common/languageServer/languageServer.ts:1]: Hm, should we treat these warnings about the SNYK_CLI_PATH?

File: src/snyk/common/platform.ts [hunk 15]
```diff
@@ -5,6 +5,10 @@ export class Platform {
     return os.platform();
   }
 
+  static getArch(): string {
+    return os.arch();
+  }
+
   static getVersion(): string {
```
Comment (bastiandoetsch) [inline src/snyk/common/platform.ts:12]: pretty sure, this is not needed anymore

File: src/snyk/common/platform.ts [hunk 16]
```diff
@@ -5,6 +5,10 @@ export class Platform {
...
     return `${os.release()}-${os.arch}`;
```
Comment (bastiandoetsch) [inline src/snyk/common/platform.ts:13]: does the arch need to be mapped? e.g. from aarch64 to arm64?

File: src/snyk/common/services/downloadService.ts [hunk 17]
```diff
@@ -81,25 +74,31 @@ export class DownloadService {
...
     }
     return false;
   }
 
-  async isLsInstalled() {
-    const lsExecutableExists = await LsExecutable.exists(this.configuration);
-    const lastUpdateDateWritten = !!this.getLastLsUpdateDate();
-    const lsChecksumWritten = !!this.getLsChecksum();
+  async isCliInstalled() {
+    const lsExecutableExists = await CliExecutable.exists(
```
Comment (bastiandoetsch) [inline src/snyk/common/services/downloadService.ts:86]: nitpick: variable name

File: src/snyk/common/services/downloadService.ts [hunk 18]
```diff
@@ -81,25 +74,31 @@ export class DownloadService {
...
+      this.extensionContext.extensionPath,
+      await this.configuration.getCliPath(),
+    );
+    const lsChecksumWritten = !!this.getCliChecksum();
 
-    return lsExecutableExists && lastUpdateDateWritten && lsChecksumWritten;
+    return lsExecutableExists && lsChecksumWritten;
```
Comment (bastiandoetsch) [inline src/snyk/common/services/downloadService.ts:92]: nitpick: variable names

File: src/test/unit/common/services/downloadService.test.ts [hunk 19]
```diff
@@ -69,9 +61,9 @@ suite('DownloadService', () => {
   test('Tries to download LS if not installed', async () => {
     configuration = {
       isAutomaticDependencyManagementEnabled: () => true,
-      getCustomCliPath: () => undefined,
-      getSnykLanguageServerPath: () => 'abc/d',
-    } as unknown as IConfiguration;
+      getCliReleaseChannel: () => 'stable',
+      getCliPath: () => Promise.resolve('path/to/cli'),
```
Comment (bastiandoetsch) [inline src/test/unit/common/services/downloadService.test.ts:65]: we should have a test that checks usage of the default value - maybe it's there, but haven't seen it yet.

File: src/test/unit/common/services/downloadService.test.ts [hunk 20]
```diff
@@ -69,9 +61,9 @@ suite('DownloadService', () => {
   test('Tries to download LS if not installed', async () => {
```
Comment (bastiandoetsch) [inline src/test/unit/common/services/downloadService.test.ts:61]: nitpick: test name is....suboptimal

File: src/test/unit/common/services/downloadService.test.ts [hunk 21]
```diff
@@ -84,11 +76,11 @@ suite('DownloadService', () => {
   test('Tries to update LS if installed', async () => {
```
Comment (bastiandoetsch) [inline src/test/unit/common/services/downloadService.test.ts:76]: nitpick: testname

File: src/test/unit/common/services/downloadService.test.ts [hunk 22]
```diff
@@ -98,112 +90,14 @@ suite('DownloadService', () => {
...
-    const downloadSpy = stub(service, 'download');
-    const updateSpy = stub(service, 'update');
-
-    await service.downloadOrUpdate();
-
-    strictEqual(downloadSpy.called, true);
-    strictEqual(updateSpy.calledOnce, false);
-  });
-
   test("Doesn't download LS if automatic dependency management disabled", async () => {
```
Comment (bastiandoetsch) [inline src/test/unit/common/services/downloadService.test.ts:93]: nitpick: name

File: src/test/unit/download/downloader.test.ts [hunk 23]
```diff
@@ -4,43 +4,40 @@ import sinon from 'sinon';
...
-import { LsExecutable } from '../../../snyk/common/languageServer/lsExecutable';
-import { IStaticLsApi, LsMetadata } from '../../../snyk/common/languageServer/staticLsApi';
+import { CliExecutable } from '../../../snyk/cli/cliExecutable';
+import { IStaticCliApi } from '../../../snyk/cli/staticCliApi';
 import { ILog } from '../../../snyk/common/logger/interfaces';
 import { LoggerMock } from '../mocks/logger.mock';
 import { windowMock } from '../mocks/window.mock';
+import { ExtensionContext } from '../../../snyk/common/vscode/extensionContext';
 
 suite('LS Downloader (LS)', () => {
```
Comment (bastiandoetsch) [inline src/test/unit/download/downloader.test.ts:14]: nitpick: name

File: src/snyk/common/configuration/configuration.ts (hunk 10 above)
Comment (ShawkyZ) [reply src/snyk/common/configuration/configuration.ts:325]:
```
  static async getPath(extensionDir: string, customPath?: string): Promise<string> {
    if (customPath) {
      return customPath;
    }

    const platform = await this.getCurrentWithArch();
    const fileName = this.getFileName(platform);
    return path.join(extensionDir, fileName);
  }
```

Comment (ShawkyZ) [reply src/snyk/common/configuration/configuration.ts:325]: the function name is confusing 
maybe it should be determineCliPath or smth

File: src/snyk/common/download/downloader.ts (hunk 11 above)
Comment (ShawkyZ) [reply src/snyk/common/download/downloader.ts:43]: https://github.com/snyk/vscode-extension/pull/551#discussion_r1822112642

File: src/snyk/common/download/downloader.ts (hunk 12 above)
Comment (ShawkyZ) [reply src/snyk/common/download/downloader.ts:44]: I left the existing logic for downloading / deleting as is. 
Currently it deletes it in the beginning if it exists. But we can change it.

File: src/snyk/common/download/downloader.ts (hunk 14 above)
Comment (ShawkyZ) [reply src/snyk/common/download/downloader.ts:97]: https://github.com/snyk/vscode-extension/pull/551/files#r1822131908

Comment (ShawkyZ) [reply src/snyk/common/download/downloader.ts:97]: I left the existing logic for downloading / deleting as is.
Currently it deletes it in the beginning if it exists. But we can change it.

File: src/snyk/common/platform.ts (hunk 16 above)
Comment (ShawkyZ) [reply src/snyk/common/platform.ts:13]: nope it returns arm64 luckily

File: src/snyk/cli/cliExecutable.ts [hunk 24]
```diff
@@ -1,37 +1,78 @@
...
-
   constructor(public readonly version: string, public readonly checksum: Checksum) {}
 
-  static getFilename(platform: CliSupportedPlatform): string {
-    return this.filenameSuffixes[platform];
-  }
-
-  static getPath(extensionDir: string, customPath?: string): string {
+  static async getPath(extensionDir: string, customPath?: string): Promise<string> {
+    // check if custom path is file
```
Comment (ShawkyZ) [inline src/snyk/cli/cliExecutable.ts:21]: TODO here

File: src/test/unit/common/services/downloadService.test.ts (hunk 19 above)
Comment (ShawkyZ) [reply src/test/unit/common/services/downloadService.test.ts:65]: test in configuration.tests.ts

File: src/snyk/common/services/downloadService.ts [hunk 25]
```diff
@@ -111,37 +110,24 @@ export class DownloadService {
...
-    return this.extensionContext.getGlobalStateValue<number>(MEMENTO_LS_LAST_UPDATE_DATE);
+  private hasCliVersionUpdated(cliVersion: string): boolean {
+    const currentVersion = this.getCliVersion();
+    return currentVersion != cliVersion;
   }
 
-  private getLsProtocolVersion() {
-    return this.extensionContext.getGlobalStateValue<number>(MEMENTO_LS_PROTOCOL_VERSION);
+  private getCliVersion(): string | undefined {
+    return this.extensionContext.getGlobalStateValue<string>(MEMENTO_CLI_VERSION);
```
Comment (bastiandoetsch) [inline src/snyk/common/services/downloadService.ts:127]: nitpick(non-blocking): We should probably clean up the old mementos

File: src/snyk/common/services/downloadService.ts [hunk 26]
```diff
@@ -111,37 +110,24 @@ export class DownloadService {
...
   }
 
-  private getLsChecksum(): number | undefined {
-    return this.extensionContext.getGlobalStateValue<number>(MEMENTO_LS_CHECKSUM);
+  private getCliChecksum(): string | undefined {
+    return this.extensionContext.getGlobalStateValue<string>(MEMENTO_CLI_CHECKSUM);
```
Comment (bastiandoetsch) [inline src/snyk/common/services/downloadService.ts:131]: nitpick(non-blocking): We should probably clean up the old mementos

File: src/snyk/common/services/downloadService.ts (hunk 25 above)
Comment (bastiandoetsch) [reply src/snyk/common/services/downloadService.ts:127]: question: where is the protocol version (of LS) saved and compared against the requiredProtocol Version? Has this change an impact on that?

File: src/test/unit/cli/cliExecutable.test.ts [hunk 27]
```diff
@@ -10,30 +11,60 @@ suite('CliExecutable', () => {
...
 
-    const winExtensionDir = `C:\\Users\\user\\.vscode\\extensions`;
-    stub.returns('win32');
+    fsStub.returns(Promise.resolve());
+    expectedCliPath = path.join(unixExtensionDir, 'snyk-alpine');
+    strictEqual(await CliExecutable.getPath(unixExtensionDir), expectedCliPath);
+    fsStub.returns(Promise.reject());
+
+    osStub.returns('win32');
+    expectedCliPath = path.join(winExtensionDir, 'snyk-win.exe');
```
Comment (bastiandoetsch) [inline src/test/unit/cli/cliExecutable.test.ts:42]: nitpick(non-blocking): I think arm64 for windows is missing in the tests? I know it works...

File: src/test/unit/download/downloader.test.ts [hunk 28]
```diff
@@ -4,75 +4,71 @@ import sinon from 'sinon';
...
 
   // noinspection DuplicatedCode
   teardown(() => {
     sinon.restore();
   });
 
-  test('Download of LS fails if platform is not supported', async () => {
-    const downloader = new Downloader(configuration, lsApi, windowMock, logger);
-    sinon.stub(LsExecutable, 'getCurrentWithArch').throws(new Error());
+  test('Download of CLI fails if platform is not supported', async () => {
```
Comment (bastiandoetsch) [inline src/test/unit/download/downloader.test.ts:48]: suggestion: test for the protocol version would be neat.

**Pull request merged**